### DISCUSSION TITLE
Change the workspace layout via the niri ipc api

### DISCUSSION THREAD

**Discussion by TheAngusMcFire** [YaLTeR/niri#708] [Ideas]
Hi everybody. 

Have you ever considered to set the layout of a workspace via the ipc api? 

I think about something like getting the layout tree via the niri msg with the window ids, titles, app_ids... in the structure of columns with the containing Windows and there height/preset, with preserved ordering the way it is currently on the workspace.

Some external application could use this feature to perform a reordering/resizing of the windows and calls niri msg with the reordered/resized layout representation.
Niri then would try to change the window locations and sizes if possible(returns error if values out of bounds, or windows missing/duplicated).

I would use it to automate workspace layouts I frequently use. e.g. work-spaces with only the messaging stuff, or work-spaces for multimedia, frequently used projects etc...

I am not sure if this could be emulated using the current niri ipc api.

**Comment by YaLTeR** [comment]
Hey, I don't want to expose the layout tree because it's hard to future proof the data structures. I am planning however to add more properties to windows, like their positions on the workspace, sizes, etc.

**Reply by druskus20** [reply]
Yes, window positions would be really good - I am assuming floating windows _probably_ come first since that would maybe imply a refactor of niri-ipc. Window positions would allow users to make widgets to indicate not only the workspaces but also the relative position of windows within a workspace. I often struggle to realize that I have more than one window on a workspace (as I mostly use fullscreen). 

Another thing that would be nice is to be able to toggle centered-column on demand. I find that with a centered column layout I don't have enough screen space to copy side by side between two windows.

**Comment by TheAngusMcFire** [comment]
Hm, good point, the layout tree is indeed hard to future proof. I figure there would be a bunch of endpoints needed to manipulate the  Layout via the IPC. Would be worth it tho, it really would be awesome tho have some way to automate workspace. Currently I open the required applications via a script but the stacking and resizing need to be done by hand. It is OK, but it is always the same :see_no_evil: 
I keep updated with the IPC changes and see if I can build the replication on my side with the IPC only.
Thanks for the reply.
